zeus/api/schemas/testcase.py

```diff
@@ -1,13 +1,132 @@
+from collections import defaultdict
 from marshmallow import Schema, fields, pre_dump
+from typing import List, Mapping
 from uuid import UUID
 
-from zeus.constants import Result
+from zeus.config import db
+from zeus.constants import Result, Status
+from zeus.models import Build, Job, Source, TestCase
 from zeus.utils.aggregation import aggregate_result
+from zeus.vcs.base import UnknownRevision
 
+from .build import BuildSchema
 from .fields import ResultField
 from .job import JobSchema
 
 
+def find_failure_origins(build: Build, test_failures: List[str]) -> Mapping[str, UUID]:
+    """
+    Attempt to find originating causes of failures.
+
+    Returns a mapping of {TestCase.hash: Job}.
+    """
+    if not test_failures:
+        return {}
+
+    source = build.source
+    repo = build.repository
+    vcs = repo.get_vcs()
+    if not vcs:
+        valid_revisions = []
+    else:
+        try:
+            valid_revisions = [
+                c.sha for c in vcs.log(limit=100, parent=source.revision_sha)
+            ]
+        except UnknownRevision:
+            valid_revisions = []
+
+    filters = [
+        Build.repository_id == build.repository_id,
+        Build.date_created <= build.date_created,
+        Build.status == Status.finished,
+        Source.id != source.id,
+        Source.date_created <= source.date_created,
+        Source.patch == None,  # NOQA
+    ]
+    if valid_revisions:
+        filters.append(Source.revision_sha.in_(valid_revisions))
+
+    # NOTE(dcramer): many of these queries ignore tenant constraints
+    # find any existing failures in the previous runs
+    # to do this we first need to find the last passing build
+    last_pass = (
```

Comment:
might have some perf issues here, but given it SHOULD be limited to ``valid_revisions`` we're safe